clients/drcachesim/tracer/tracer.cpp

```diff
@@ -413,13 +422,29 @@ memtrace(void *drcontext, bool skip_size_cap)
                                               dr_get_thread_id(drcontext));
     pipe_start = data->buf_base;
     pipe_end = pipe_start;
-    if (!skip_size_cap && op_max_trace_size.get_value() > 0 &&
-        data->bytes_written > op_max_trace_size.get_value()) {
+    if (!skip_size_cap &&
+        (data->output_disabled ||
+         (((op_max_trace_size.get_value() > 0 &&
+            data->bytes_written > op_max_trace_size.get_value()) ||
+           is_beyond_global_max())))) {
         /* We don't guarantee to match the limit exactly so we allow one buffer
          * beyond.  We also don't put much effort into reducing overhead once
          * beyond the limit: we still instrument and come here.
          */
         do_write = false;
+        if (!data->output_disabled && is_beyond_global_max()) {
```

Comment:
Instead of using an additional `output_disabled` var, we can maybe just check `notify_once>0`? This if-block will be executed at-most once for each thread then, but that's okay right?

However, if the `output_disabled` var was added to make the state explicit, then let's keep it.

Reply:
The local var was probably a misguided attempt to reduce atomics on every trip once over the threshold, but this output function is on a slowpath in any case and one barrier is probably fine.  Yes I am thinking of either going with an atomic read of notify_once, which should prevent reordering the store above it (it is an acquire operation in DR: we should make that explicit in the docs), or going with std::atomic<bool>.

Reply:
Went with atomic load.